### ISSUE TITLE
Colour cast if PPG demosaicing is used with OpenCL acceleration

### ISSUE THREAD

**Issue by mkunzBGC** [darktable-org/darktable#18042]
### Describe the bug

If I use OpenCL acceleration in connection with PPG demosaicing, my photos get a strong green colour cast:
![ppgDemosaicOpenCl](https://github.com/user-attachments/assets/3e1cb2d0-472e-402f-88be-cc9db3dde6a1)

The image in the navigation pane looks fine (as far as I understand that image is processed by the CPU).

Turning off OpenCL acceleration removes the colour cast:
![ppgDemosaicCpu](https://github.com/user-attachments/assets/0bb5a88e-6afa-4f36-a28a-9fd569326b9a)

... as does switching to any other demosaic algorithm, e.g. RCD, with OpenCL enabled:
![rcdDemosaicOpenCl](https://github.com/user-attachments/assets/09b03256-8200-4897-994c-3074ad916319)

If OpenCL is enabled and I use PPG, the exported files have the colour cast as well, see
[KNZ_1421PpgDemosaicOpenCl](https://github.com/user-attachments/assets/cffc42e9-bdb7-48f8-82ac-3fc79dbd12fc).

The sample raw image plus sidecar file can be downloaded [here](https://nc.imfus.de/index.php/s/4AmLnXCQXHQoSjK) .

### Steps to reproduce

1. With OpenCL enabled, open the provided sample raw image in darkroom.
As long as PPG is used for demosaicing, the main image will have a strong green colour cast.

### Expected behavior

The colours in the main view should be the same as in the navigation pane. They should also not change when OpenCL is enabled/disabled.

### Logfile | Screenshot | Screencast

I am attaching a log file created by executing darktable with ` -d common -d opencl`. During that session I opened my sample raw image in question darkroom and then closed darktable.
[darktableLog.txt](https://github.com/user-attachments/files/18222462/darktableLog.txt)

### Where did you obtain darktable from?

downloaded from www.darktable.org

### darktable version

4.8.1

### What OS are you using?

Windows

### What is the version of your OS?

Windows 11 Pro, 24H2

### Describe your system?

Processor: Intel(R) Core(TM) i9-14900KF (32 CPUs), ~3.2GHz
Memory: 65536MB RAM
Video Card: AMD Radeon RX 7800 XT
Video Memory: 16152 MB
Driver Version (according to device manager): 32.0.12019.1028
Driver Date: 2024-10-11

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

AMD Radeon RX 7800 XT, 16 GB, Driver Version (according to device manager): 32.0.12019.1028

**Comment by jenshannoschwalm** [comment]
Tested and couldn't reproduce with raw and xmp on 5.0.

There has been a major overhaul of the demosaicer code in 5.0, please try 5.0 and report back if issue is persisting. From the logs i have no clue.

**Comment by mkunzBGC** [comment]
Thanks for looking into this.

I downloaded darktable 5.0.0 ([this link](https://github.com/darktable-org/darktable/releases/download/release-5.0.0/darktable-5.0.0-win64.exe)) and tested again, with the same results as 4.8.1:

- The XMP file linked in my original post yields an image with a green colour cast if OpenCL is enabled. 
- A duplicate of the same image with the history reset (and module order v5.0 RAW) exhibits the same problem.
- If OpenCL is disabled and darktable restarted, both images look fine.

The JPG file exported while OpenCL is enabled has the same problem ([KNZ_1421_PpgOpenCl5 0 0](https://github.com/user-attachments/assets/a5d9d966-7d6e-43ee-b6cf-f387f86a990d)). What I notice in this full size export is a strong Moiré pattern - maybe this gives a hint as to what is going wrong?

**Comment by jenshannoschwalm** [comment]
It's only happening with PPG demosaicer ?

Other demosaicers work fine?

**Comment by mkunzBGC** [comment]
Yes, none of AMaZE, VNG4, RCD and LMMSE exhibit the problem, only PPG. Behaviour is the same in 4.8 and 5.0.

Am 28. Dezember 2024 08:51:27 MEZ schrieb Hanno Schwalm ***@***.***>:

**Comment by jenshannoschwalm** [comment]
Ok and thanks for reporting. PPG is not used often nowadays so likely got unnoticed.

What about other modules? All good?

**Comment by jenshannoschwalm** [comment]
@piratenpanda Hi Benjamin, i guess you are on AMD? Can you confirm this?

**Comment by piratenpanda** [comment]
from the log I see 3628.0 as opencl version. I doubt that this has been fixed with newer drivers, but it might be worth a shot to upgrade to latest rocm.

**Comment by jenshannoschwalm** [comment]
> from the log I see 3628.0 as opencl version.

Ahh. I missed that.

**Comment by europlatus** [comment]
Confirmed issue is present on 5.0.1 with PPG as demosaicer and with AMD openCL active.
